compiler/rustc_parse/src/parser/item.rs

```diff
@@ -247,9 +247,37 @@ impl<'a> Parser<'a> {
             (ident, ItemKind::Static(ty, m, expr))
         } else if let Const::Yes(const_span) = self.parse_constness() {
             // CONST ITEM
-            self.recover_const_mut(const_span);
-            let (ident, ty, expr) = self.parse_item_global(None)?;
-            (ident, ItemKind::Const(def(), ty, expr))
+            if self.token.is_keyword(kw::Impl) {
+                // recover from `const impl`, suggest `impl const`
+                let impl_span = self.token.span;
+                let mut err = self.expected_ident_found();
+                let mut impl_info = self.parse_item_impl(attrs, def())?;
+                match impl_info.1 {
+                    // only try to recover if this is implementing a trait for a type
+                    ItemKind::Impl { of_trait: Some(ref trai), ref mut constness, .. } => {
+                        *constness = Const::Yes(const_span);
+
+                        let before_trait = trai.path.span.shrink_to_lo();
+                        let const_up_to_impl = const_span.with_hi(impl_span.lo());
+                        err.multipart_suggestion(
+                            "you might have meant to write a const trait impl",
+                            vec![
+                                (const_up_to_impl, "".to_owned()),
+                                (before_trait, "const ".to_owned()),
+                            ],
+                            Applicability::MaybeIncorrect,
+                        )
+                        .emit();
+                    }
+                    ItemKind::Impl { .. } => return Err(err),
+                    _ => unreachable!(),
+                }
+                impl_info
```

Comment:
Could you move this code to a separate function in `parser/diagnostics.rs`?
That's where diagnostic or recovery code lives, so it's separated from the "reference parser" code describing the language.

Reply:
Oh I only just saw that you asked to move it to `diagnostics.rs`; I just moved it under `recover_const_mut` which is still in `parser/item.rs`, there's other `recover_*` functions there as well.

Reply:
>`parser/item.rs`, there's other `recover_*` functions there as well.

(The other function will also need to move to `diagnostics.rs`, but not in this PR.)